### ISSUE TITLE
Decimal Deserialization Doesn't Maintain Accuracy

### ISSUE THREAD

**Issue by sdelano** [pydantic/pydantic#6295]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

While using FastAPI 0.100.0b1 and Pydantic 2.0b3 I've noticed that decimals deserialized from JSON no longer maintain the accuracy as stated in the [documentation](https://docs.pydantic.dev/dev-v2/usage/types/number_types/#validation-of-numeric-types).

>  Pydantic attempts to convert the value to a string, then passes the string to Decimal(v).

It's clear from simple testing that Pydantic v2 no longer does this, and the behavior matches what would happen if you created a decimal from a float, like so:

```python
decimal.Decimal(48.717983)
```

where the documentation (and v1 behavior) states that the following should happen:

```python
decimal.Decimal("48.717983")
```

This behavior results in decimal fields that do not accurately match the intended value.

### Example Code

```Python
#!/usr/bin/env python

import decimal
import pydantic

data = '{"number":48.717983}'

class Test(pydantic.BaseModel):
    number: decimal.Decimal

parsed = Test.model_validate_json(data)  # use parse_raw for v1
expected = decimal.Decimal("48.717983")
assert parsed.number == expected
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0b3
        pydantic-core version: 0.42.0 release build profile
                 install path: /Users/stephen/code/pydantic/pydantic
               python version: 3.11.4 (main, Jun 13 2023, 16:13:25) [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-13.4.1-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```
```


Selected Assignee: @dmontagu

**Comment by dmontagu** [comment]
For what it’s worth, you can use a string instead of a number in the JSON to resolve this, but I understand if you can’t control that and agree it’s worth a fix.

@davidhewitt do you know your way around arbitrary precision floating points from JSON using serde? I remember trying to get arbitrary size ints working to handle ints outside the range of i64, but it was a little bit over my head, and I’m not sure if there’s something analogous for floats anyway..

**Comment by sdelano** [comment]
Thanks for picking this up @dmontagu! We're keeping an eye on this as we pass around decimals frequently in our system to be used within financial calculations. Before we added Pydantic to our toolset, we were very intentional about the way that we handled the serialization and de-serialization of these types. One of the big (among many) positives of adopting Pydantic is that it took the same approach as we did, but out of the box without any extra work.

It's unclear if that approach is changing in Pydantic v2, as the docs conflict themselves.

[v2 Number Types](https://docs.pydantic.dev/dev-v2/usage/types/number_types/#validation-of-numeric-types):

> Pydantic attempts to convert the value to a string, then passes the string to Decimal(v).

[v2 Migration Guide](https://docs.pydantic.dev/dev-v2/migration/#validator-behavior-changes)

> int, float, and decimal values are no longer coerced to strings

If anything, you've at least given us a direction that we can look to pursue if we need to implement our own serialization.

**Comment by dmontagu** [comment]
@sdelano

> int, float, and decimal values are no longer coerced to strings

This is a reference to the idea that if you have a `str` field, you can't pass a Decimal:
```python
from decimal import Decimal

from pydantic import BaseModel


class X:
    def __str__(self):
        return '1.1'


class Model(BaseModel):
    x: Decimal


Model(x=Decimal('1.1'))
"""
x
  Input should be a valid string [type=string_type, input_value=Decimal('1.1'), input_type=Decimal]
    For further information visit https://errors.pydantic.dev/2.0.0/v/string_type
"""
```

> Pydantic attempts to convert the value to a string, then passes the string to Decimal(v).

This is saying that if you give a value for a decimal, it will have `str` called on it then `Decimal`. However, in v2, we are a bit more strict about what the inputs are allowed to be (must be a `str`/`int`/`float`/`Decimal`, not an arbitrary value).

That said, I've spoken with @samuelcolvin about this, and I think our conclusion is that we just need to add better support for Decimal directly to `pydantic-core`, which would ensure that even when parsing from JSON you will get perfect precision out of numeric types. (Apparently @samuelcolvin already has a trick for this from the `speedate` library so it shouldn't be too difficult.)

That said, we've _just_ released v2.0 and didn't get a fix in for this, but I think this is a great candidate for v2.1 (or perhaps a patch release even sooner). We'll keep you updated here.
